### DISCUSSION TITLE
Capture HL7V2 Segment via Attachment Handler

### DISCUSSION THREAD

**Discussion by JJWest12** [nextgenhealthcare/connect#5964] [Q&A]
I am trying to use a JavaScript attachment handler to return only the MSH segment of an HL7V2 message. My understanding is that in the attachment handler you first need to serialize the message, which I believe I do correctly below. The issue I am running in to with this though is that my return is saying "undefined". Does anyone know what I am missing here?

    var dataType = 'HL7V2';
    var serializationProperties = SerializerFactory.getDefaultSerializationProperties(dataType);
    var deserializationProperties = SerializerFactory.getDefaultDeserializationProperties(dataType);
    var serializer = SerializerFactory.getSerializer(dataType, serializationProperties, deserializationProperties);

    return message['MSH'];

**Comment by jonbartels** [comment]
You never call `serializer` to do something with `message`.

**Reply by JJWest12** [reply]
I actually figured out what I needed to do to get what I needed in to an attachment. The issue I am now having is trying to add the attachment back to the message on the destination that sends to my next channel. I have been looking into the user guide and see using "${ATTACH..." but I cannot get it to work out.

My destination is sending off something like this below, with 'b' being the attachment ID that I map in the source transformer, but the message literally just has the string "${ATTACH:${b}}" rather than the attachment.

    ${message.encodedData}${ATTACH:${b}}

**Reply by jonbartels** [reply]
https://forums.mirthproject.io/articles/175046-using-attachments is another good reference.

Try this: Build `b` as the STRING `${ATTACH:your-uuid-here}` then make your template `${message.encodedData}${b}`

**Reply by tonygermano** [reply]
You shouldn't need to build the attachment token string at all. Share your code for how you are creating the attachment. Your original post had to do with serializing an hl7 message in the attachment handler, and now you're talking about doing something else in the source transformer.

Depending on your hl7 you might be able to use a regex attachment handler and skip all of this javascript stuff.

**Reply by JJWest12** [reply]
I am super late on this.. I confirmed that the code in my attachment handler is doing what it is I want it to do. The issue is getting that attachment back on to the message before sending it off the the endpoint in the destinations. From what I can tell, you need the format "${ATTACH:___" in the message as well as "Reattach Attachments" sent to Yes in the destination itself. So I have code in my source transformer to manually create that "${ATTACH:___" string if an attachment is present (not every message in this channel will have an attachment created).

I created this if/else because i didn't want "$ATTACH:___" present on my messages if there was no attachment created for a particular message. Is there a different way of getting the attachment added back on to the message prior to sending to the endpoint other than doing it this way?

if(attachmentIdentifiers.length == 0){
	channelMap.put('attachmentString', '')
}
else{
	channelMap.put('attachmentString', '${ATTACH:' + firstAttachmentId + '}')
}

**Reply by tonygermano** [reply]
I still think you're making this way more complicated than it needs to be. Share **all** of your attachment related code. You don't need to construct the attachment token yourself. The attachment token *should* be in the message if anything was replaced. If nothing was replaced, then you don't need a token. There's usually no reason to put the token in a channelMap variable. Regex is probably still a simpler solution.
